### PR TITLE
cilium: enable bpf masquerading

### PR DESCRIPTION
**Pull request by 3u13r**
### Proposed change(s)

- Enable Cilium bpf masquerading in order to be able to configure masquerading through a configmap (see: https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based)

### Additional info

This also enables full eBPF host routing:
Before:
```
root@fedora:/home/cilium# cilium status
KVStore:                 Ok   Disabled
Kubernetes:              Ok   1.28 (v1.28.3) [linux/amd64]
Kubernetes APIs:         ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "cilium/v2alpha1::CiliumEndpointSlice", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:    Strict   [eth0 10.9.0.6]
Host firewall:           Disabled
CNI Chaining:            none
Cilium:                  Ok   1.12.1 (v1.12.1-d465085df1)
NodeMonitor:             Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:    Ok
IPAM:                    IPv4: 19/254 allocated from 10.244.0.0/24,
BandwidthManager:        Disabled
Host Routing:            Legacy
Masquerading:            IPTables [IPv4: Enabled, IPv6: Disabled]
Controller Status:       89/89 healthy
Proxy Status:            No managed proxy redirect
Global Identity Range:   min 256, max 65535
Hubble:                  Ok              Current/Max Flows: 3040/4095 (74.24%), Flows/s: 28.02   Metrics: Disabled
Encryption:              Wireguard       [cilium_wg0 (Pubkey: Y9fUb7bOQKvO9HgMxgttq0AOwnp10fVohMqAFrf7rSc=, Port: 51871, Peers: 0)]
Cluster health:          1/1 reachable   (2023-12-15T13:23:03Z)
```

After:
```
root@fedora:/home/cilium# cilium status
KVStore:                 Ok   Disabled
Kubernetes:              Ok   1.28 (v1.28.4) [linux/amd64]
Kubernetes APIs:         ["EndpointSliceOrEndpoint", "cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "cilium/v2alpha1::CiliumCIDRGroup", "cilium/v2alpha1::CiliumEndpointSlice", "core/v1::Namespace", "core/v1::Pods", "core/v1::Service", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:    Strict   [ens3 192.168.178.3 (Direct Routing), cilium_wg0 ]
Host firewall:           Disabled
CNI Chaining:            none
Cilium:                  Ok   1.15.0-pre.2 (v1.15.0-pre.2-0f91509967)
NodeMonitor:             Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:    Ok
IPAM:                    IPv4: 9/254 allocated from 10.10.3.0/24,
IPv4 BIG TCP:            Disabled
IPv6 BIG TCP:            Disabled
BandwidthManager:        Disabled
Host Routing:            BPF
Masquerading:            BPF (ip-masq-agent)   [ens3, cilium_wg0]   10.10.0.0/16 [IPv4: Enabled, IPv6: Disabled]
Controller Status:       47/47 healthy
Proxy Status:            No managed proxy redirect
Global Identity Range:   min 256, max 65535
Hubble:                  Ok              Current/Max Flows: 4095/4095 (100.00%), Flows/s: 47.66   Metrics: Disabled
Encryption:              Wireguard       [NodeEncryption: Enabled, cilium_wg0 (Pubkey: NQsZTCE7Z3/MA7jPNdRdaOPttZkVF+aUaijiEdd+2ww=, Port: 51871, Peers: 3)]
Cluster health:          4/4 reachable   (2023-12-15T13:28:37Z)
Modules Health:          Stopped(0) Degraded(0) OK(6) Unknown(1)
```

### Checklist

- [x] Add labels (e.g., for changelog category)
- [x] Link to Milestone

### CODE REVIEW

**Comment by 3u13r** [comment]
upgrade test, azure, 3:1: https://github.com/edgelesssys/constellation/actions/runs/7219396872
upgrade test, gcp, 3:1: https://github.com/edgelesssys/constellation/actions/runs/7219399602

**Review by burgerdev — APPROVED** [review]
Thanks for sending the PR! In order for us to be able to fine-tune the masquerading for on-prem IPs in a VPN setup, we might want to also set `ipMasqAgent.enabled=true`. The default config seems to include all private IP spaces, but maybe that's too much?

**Comment by 3u13r** [comment]
> The default config seems to include all private IP spaces, but maybe that's too much?

That's a fair point. Ideally we enable that, but then we should also configure it. I think we have the necessary information already, i.e. normally only node IPs shouldn't be masqued. On GCP we simply add the ip cidr.

**Comment by 3u13r** [comment]
New upgrade tests:
upgrade test, azure, 3:1: https://github.com/edgelesssys/constellation/actions/runs/7222748017
upgrade test, gcp, 3:1: https://github.com/edgelesssys/constellation/actions/runs/7222757123

File: internal/constellation/helm/overrides.go [hunk 1]
```diff
@@ -59,6 +59,12 @@ func extraCiliumValues(provider cloudprovider.Provider, conformanceMode bool, ou
 	extraVals["encryption"] = map[string]any{
 		"strictMode": strictMode,
 	}
+	extraVals["ipMasqAgent"] = map[string]any{
+		"config": map[string]any{
+			"nonMasqueradeCIDRs": []string{output.IPCidrNode},
```
Comment (burgerdev) [inline internal/constellation/helm/overrides.go:64]: Shouldn't this contain the pod network, too?

Comment (3u13r) [reply internal/constellation/helm/overrides.go:64]: Only on GCP right? There it is enabled through ipv4NativeRoutingCIDR (I manually checked this with tcpdump).

Comment (3u13r) [reply internal/constellation/helm/overrides.go:64]: Also on the node who has the 10.244.1.0/24 pod cidr I get the following iptables on Azure on the main:
```
root@fedora:/home/cilium# iptables -vL -t nat 
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  114  9844 CILIUM_PRE_nat  all  --  any    any     anywhere             anywhere             /* cilium-feeder: CILIUM_PRE_nat */

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  220 13034 CILIUM_OUTPUT_nat  all  --  any    any     anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT_nat */

Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  291 19192 CILIUM_POST_nat  all  --  any    any     anywhere             anywhere             /* cilium-feeder: CILIUM_POST_nat */

Chain CILIUM_OUTPUT_nat (1 references)
 pkts bytes target     prot opt in     out     source               destination         

Chain CILIUM_POST_nat (1 references)
 pkts bytes target     prot opt in     out     source               destination         
   73  4627 MASQUERADE  all  --  any    eth0    anywhere            !10.244.1.0/24        /* cilium masquerade non-cluster */
    0     0 ACCEPT     all  --  any    any     anywhere             anywhere             mark match 0xa00/0xe00 /* exclude proxy return traffic from masquerade */
    0     0 SNAT       all  --  any    cilium_host !10.244.1.0/24       !10.244.1.0/24        /* cilium host->cluster masquerade */ to:10.244.1.47
    0     0 SNAT       all  --  any    lxc+    localhost            anywhere             /* cilium host->cluster from 127.0.0.1 masquerade */ to:10.244.1.47

Chain CILIUM_PRE_nat (1 references)
 pkts bytes target     prot opt in     out     source               destination         

Chain KUBE-KUBELET-CANARY (0 references)
 pkts bytes target     prot opt in     out     source               destination     
```

**Pull request merged**